Fixed crit calculation and reduced time complexity

psdecabooter opened the pull request:
Closes #15

Abtin-Molavi reviewed (commented):
This looks good to me. Have you checked the correctness on some examples?

psdecabooter commented:
I have done correctness on some examples. However, I think it would be a good idea for some regression tests to be made. Especially because we are doing a rewrite.

psdecabooter commented:
I just added another commit, all it does was switch from reversed(enumerate(gates)) to actually just a for loop moving an index backwards range(len(gates)-1,-1,-1). It says I changed a lot because I accidentally formatted the file too. We might want to make a github workflow which will enforce some linting rules.
